proposals/0360-opaque-result-types-with-availability.md

```diff
@@ -87,18 +87,123 @@ This is a dead-end for the library author although SE-0244 states that it should
 
 ## Proposed solution
 
-To bridge this usability gap I propose to relax same-type restriction for `return`s inside of availability conditions as follows:
+To bridge this usability gap I propose to relax same-type restriction for functions with `if #available` conditions if they meet all of the following conditions.
 
-A function returning an opaque result type is allowed to return values of different concrete types from conditionally available `if #available`  branches without any other dynamic conditions, if and only if, all universally available `return` statements in its body return a value of the same concrete type `T`. All returned values regardless of their location must meet all of the constraints stated on the opaque type.
+1. There is at least one universally (unconditionally) available `return` in a function. All universally available `return`s are required to produce the same type.
+
+2. `if #available` is a top-level statement in a function and *is not* preceded by any block that can return.
+
+  ```swift
+  func test() -> some Shape {
+    if #available(macOS 100, *) { ✅
+      return Rectangle()
+    }
+
+    return self
+  }
+  ```
+  
+  The following is incorrect because `if #available` is preceded by a dynamic condition that returns:
+  
+  ```swift
+  func test() -> some Shape {
+    guard let x = <opt-value> else {
+      return ...
+    }
+    
+    if #available(macOS 100, *) { ❌
+      return Rectangle()
+    }
+
+    return self
+  }
+  ```
+
+  `if #available` appears inside of a loop:
+
+  ```swift
+  func test() -> some Shape {
+    for ... {
+      if #available(macOS 100, *) { ❌
+        return Rectangle()
+      }
+    }
+    return self
+  }
+  ```
+
+3. `if #available` has to terminate with a `return`, and all `return` statements inside of a particular `if #available` condition have to produce the same type, which, at the same time, is allowed be different from top-level `return`(s) or `return`s nested in other conditions regardless of their kind.
```

Comment:
```suggestion
All `return` statements outside of unconditional availability clauses must return the same type as each other, and this type must be as available as the containing function.

All `return` statements within a given unconditional availability clause must return the same type as each other, and this type must be as available as the `#available` condition of the clause.  This type need not be the same type returned by any `return` statement outside of the clause.

There must be at least one `return` statement in the containing function.  If there are no `return` statements outside of unconditional availability clauses, then at least one of the return types within unconditional availability clauses must be as available as the containing function.

Dynamically, the return type of the containing function is:
- the return type of `return` statements in the first unconditional availability clause whose condition is dynamically satisfied, or if none are satisfied then
- the return type of `return` statements outside of all unconditional availability clauses, or if there are no such statements then
- the return type of `return` statements in the first unconditional availability clause that is as available as the containing function.
```

This is slightly more general than before because it allows code like `if #available { return } else { throw ... }`.  Note that that's allowed today as long as the return type is as available as the function.  Basically, the way I've written it, we make a decision tree as long as we syntactically match the pattern, and then maybe the decision tree can be trivialized because every branch produces the same type.